File: src/git.ts
import { spawn, spawnSync } from "node:child_process";

export type GitResult = {
  ok: boolean;
  stdout: string;
  stderr: string;
  status: number | null;
};

export function runGit(cwd: string, args: readonly string[]): GitResult {
  const result = spawnSync("git", ["-C", cwd, ...args], {
    encoding: "utf8",
    maxBuffer: 1024 * 1024 * 20,
  });

  return {
    ok: result.status === 0,
    stdout: result.stdout ?? "",
    stderr: result.stderr ?? result.error?.message ?? "",
    status: result.status,
  };
}

export function getTopLevel(repoPath: string): GitResult {
  return runGit(repoPath, ["rev-parse", "--show-toplevel"]);
}

export function getShortStatus(repoPath: string): GitResult {
  return runGit(repoPath, ["status", "--short", "--branch", "--ahead-behind"]);
}

export function getWorktreeList(repoPath: string): GitResult {
  return runGit(repoPath, ["worktree", "list", "--porcelain"]);
}

export function fastForward(repoPath: string): GitResult {
  return runGit(repoPath, ["merge", "--ff-only"]);
}

export function parseWorktreePaths(porcelain: string): string[] {
  const paths: string[] = [];

  for (const line of porcelain.split(/\r?\n/)) {
    if (line.startsWith("worktree ")) {
      paths.push(line.slice("worktree ".length));
    }
  }

  return paths;
}

export function fetchAll(repoPath: string, timeoutMs = 30_000): Promise<GitResult> {
  return new Promise((resolve) => {
    const child = spawn("git", ["-C", repoPath, "fetch", "--all"], {
      stdio: ["ignore", "pipe", "pipe"],
    });

    let stdout = "";
    let stderr = "";

    child.stdout?.on("data", (chunk: Buffer) => {
      stdout += chunk.toString("utf8");
    });

    child.stderr?.on("data", (chunk: Buffer) => {
      stderr += chunk.toString("utf8");
    });

    const timer = setTimeout(() => {
      child.kill("SIGKILL");
    }, timeoutMs);

    child.on("close", (status) => {
      clearTimeout(timer);
      resolve({
        ok: status === 0,
        stdout,
        stderr,
        status,
      });
    });

    child.on("error", (error) => {
      clearTimeout(timer);
      resolve({
        ok: false,
        stdout,
        stderr: error.message,
        status: null,
      });
    });
  });
}

File: internal/git/git.go
package git

import (
	"bytes"
	"context"
	"os/exec"
	"strings"
	"time"
)

type Result struct {
	OK     bool
	Stdout string
	Stderr string
	Status int
}

func Run(cwd string, args ...string) Result {
	fullArgs := append([]string{"-C", cwd}, args...)
	cmd := exec.Command("git", fullArgs...)
	var stdout bytes.Buffer
	var stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	err := cmd.Run()
	status := 0
	if err != nil {
		status = -1
		if exitErr, ok := err.(*exec.ExitError); ok {
			status = exitErr.ExitCode()
		} else if stderr.Len() == 0 {
			stderr.WriteString(err.Error())
		}
	}

	return Result{
		OK:     err == nil,
		Stdout: stdout.String(),
		Stderr: stderr.String(),
		Status: status,
	}
}

func TopLevel(repoPath string) Result {
	return Run(repoPath, "rev-parse", "--show-toplevel")
}

func ShortStatus(repoPath string) Result {
	return Run(repoPath, "status", "--short", "--branch", "--ahead-behind")
}

func WorktreeList(repoPath string) Result {
	return Run(repoPath, "worktree", "list", "--porcelain")
}

func FastForward(repoPath string) Result {
	return Run(repoPath, "merge", "--ff-only")
}

func FetchAll(repoPath string, timeout time.Duration) Result {
	if timeout == 0 {
		timeout = 30 * time.Second
	}

	ctx, cancel := context.WithTimeout(context.Background(), timeout)
	defer cancel()

	cmd := exec.CommandContext(ctx, "git", "-C", repoPath, "fetch", "--all")
	var stdout bytes.Buffer
	var stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	err := cmd.Run()
	status := 0
	if err != nil {
		status = -1
		if ctx.Err() == context.DeadlineExceeded {
			stderr.WriteString("git fetch timed out")
		} else if exitErr, ok := err.(*exec.ExitError); ok {
			status = exitErr.ExitCode()
		} else if stderr.Len() == 0 {
			stderr.WriteString(err.Error())
		}
	}

	return Result{
		OK:     err == nil,
		Stdout: stdout.String(),
		Stderr: stderr.String(),
		Status: status,
	}
}

func ParseWorktreePaths(porcelain string) []string {
	paths := []string{}
	for _, line := range strings.Split(strings.ReplaceAll(porcelain, "\r\n", "\n"), "\n") {
		if strings.HasPrefix(line, "worktree ") {
			paths = append(paths, strings.TrimPrefix(line, "worktree "))
		}
	}
	return paths
}
